src/MonitorProfiler/Utilities/NameCache.cpp

```diff
@@ -0,0 +1,223 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#include "NameCache.h"
+#include "macros.h"
+#include "corhlpr.h"
+
+const tstring NameCache::CompositeClassName = _T("_CompositeClass_");
+const tstring NameCache::ArrayClassName = _T("_ArrayClass_");
+const tstring NameCache::UnknownName = _T("_Unknown_");
+const tstring NameCache::ModuleSeparator = _T("!");
+const tstring NameCache::FunctionSeperator = _T(".");
+const tstring NameCache::NestedSeparator = _T("+");
+const tstring NameCache::GenericBegin = _T("<");
+const tstring NameCache::GenericSeparator = _T(",");
+const tstring NameCache::GenericEnd = _T(">");
+
+bool NameCache::GetFunctionData(FunctionID id, std::shared_ptr<FunctionData>& data)
+{
+    return GetData(_functionNames, id, data);
+}
+bool NameCache::GetClassData(ClassID id, std::shared_ptr<ClassData>& data)
+{
+    return GetData(_classNames, id, data);
+}
+bool NameCache::GetModuleData(ModuleID id, std::shared_ptr<ModuleData>& data)
+{
+    return GetData(_moduleNames, id, data);
+}
+bool NameCache::GetTokenData(ModuleID modId, mdTypeDef token, std::shared_ptr<TokenData>& data)
+{
+    auto const& it = _names.find(std::make_pair(modId, token));
+
+    if (it != _names.end())
+    {
+        data = it->second;
+        return true;
+    }
+    return false;
+}
+
+void NameCache::AddModuleData(ModuleID moduleId, tstring&& name)
+{
+    _moduleNames.emplace(moduleId, std::make_shared<ModuleData>(std::move(name)));
+}
+
+HRESULT NameCache::GetFullyQualifiedName(FunctionID id, tstring& name)
+{
+    HRESULT hr;
+
+    std::shared_ptr<FunctionData> functionData;
+    if (GetFunctionData(id, functionData))
+    {
+        std::shared_ptr<ClassData> classData;
+        if (functionData->GetClass() != 0)
+        {
+            IfFailRet(GetFullyQualifiedClassName(functionData->GetClass(), name));
+        }
+        else
+        {
+            IfFailRet(GetFullyQualifiedClassName(functionData->GetModuleId(), functionData->GetClassToken(), name));
+        }
+
+        name += FunctionSeperator + functionData->GetName();
+
+        for (int i = 0; i < functionData->GetTypeArgs().size(); i++)
+        {
+            if (i == 0)
+            {
+                name += GenericBegin;
+            }
+
+            tstring genericParamName;
+            IfFailRet(GetFullyQualifiedClassName(functionData->GetTypeArgs()[i], genericParamName));
+            name += genericParamName;
+
+            if (i == (functionData->GetTypeArgs().size() - 1))
+            {
+                name += GenericEnd;
+            }
+            else
+            {
+                name += GenericSeparator;
+            }
+        }
+
+        std::shared_ptr<ModuleData> moduleData;
+        if (GetModuleData(functionData->GetModuleId(), moduleData))
+        {
+            name = moduleData->GetName() + ModuleSeparator + name;
```

Comment:
Just out of curiosity, why do we do this last? It seems like everything else builds by appending the items in order, so I wasn't sure why this didn't follow that pattern by being checked first.

Reply:
This was a side effect of two things:
- I kept changing whether the module id should be added in the Function or the Class api
- I would ultimately like some configurability around this. I.e. making the module name optional.

I can change the order for now.

Reply:
There was another reason here which is my functions are a bit inconsistent; some append data others replace it. I will leave this in place for now and will add some TODO's around cleanup.